gateway/service/src/main/java/org/eclipse/ditto/gateway/service/endpoints/actors/CheckPermissionsActor.java

```diff
@@ -0,0 +1,417 @@
+        final Map<String, Map<String, PermissionCheckWrapper>> groupedByEntityId =
+                groupByEntityAndPolicyResource(command);
+
+        CompletableFuture<Void> allPolicyRetrievals = CompletableFuture.allOf(
+                groupedByEntityId.keySet().stream()
+                        .map(entityId -> retrieveOrSetPolicyId(entityId, groupedByEntityId.get(entityId), command))
+                        .toArray(CompletableFuture[]::new)
+        );
+
+        allPolicyRetrievals.thenRun(
+                        () -> handlePolicyRetrievalCompletion(command, permissionResults, groupedByEntityId))
+                .exceptionally(ex -> handleFailure(ex, command));
+    }
+
+    /**
+     * Initializes the permission results map with 'false' for each permission check.
+     *
+     * @param command the {@link CheckPermissions} command containing the permission checks.
+     * @return a map with the permission check results, initialized to {@code false}.
+     */
+    private Map<String, Boolean> initializePermissionResults(final CheckPermissions command) {
+        return command.getPermissionChecks().keySet().stream()
+                .collect(Collectors.toMap(
+                        check -> check,
+                        check -> false,
+                        (oldValue, newValue) -> oldValue,
+                        LinkedHashMap::new));
+    }
+
+    /**
+     * Groups permission checks by entity ID and policy resource.
+     *
+     * @param command the {@link CheckPermissions} command.
+     * @return a map grouped by entity ID, containing permission check wrappers.
+     */
+    private Map<String, Map<String, PermissionCheckWrapper>> groupByEntityAndPolicyResource(
+            final CheckPermissions command) {
+        return command.getPermissionChecks().entrySet().stream()
+                .collect(Collectors.groupingBy(
+                        entry -> entry.getValue().getEntityId(),
+                        Collectors.toMap(
+                                Map.Entry::getKey,
+                                entry -> new PermissionCheckWrapper(entry.getValue())
+                        )
+                ));
+    }
+
+    /**
+     * Retrieves or sets the {@link PolicyId} for a given entity.
+     * <p>
+     * If the entity is a policy resource, it directly sets the {@link PolicyId} for the corresponding wrappers.
+     * If the entity is not a policy resource, it retrieves the {@link PolicyId} from the entity using the
+     *
+     * @param entityId the ID of the entity for which the {@link PolicyId} is being retrieved or set.
+     * @param wrappers the wrappers containing the permission checks for the entity.
+     * @param command the {@link CheckPermissions} command containing the headers for the request.
+     * @return a {@link CompletableFuture} that completes when the {@link PolicyId} is set for all wrappers.
+     */
+    private CompletableFuture<Void> retrieveOrSetPolicyId(final String entityId,
```

Comment:
This is problematic as this relies on a mutable `PermissionCheckWrapper` class.
In order to avoid concurrency issues, we must get rid of the mutability of `PermissionCheckWrapper`.